sysboundary/ionosphere.cpp

```diff
@@ -555,75 +554,169 @@ namespace SBC {
+                     return 0.0;
+                  }
+                  return retval / nCells;
+               }
+            case 2:
+               if (  (technicalGrid.get(i,j,k-1)->SOLVE & compute::BZ) == compute::BZ
+                  && (technicalGrid.get(i,j,k+1)->SOLVE & compute::BZ) == compute::BZ
+               ) {
+                  return 0.5 * (bGrid->get(i,j,k-1)->at(fsgrids::bfield::PERBZ) + bGrid->get(i,j,k+1)->at(fsgrids::bfield::PERBZ));
+               } else if ((technicalGrid.get(i,j,k-1)->SOLVE & compute::BZ) == compute::BZ) {
+                  return bGrid->get(i,j,k-1)->at(fsgrids::bfield::PERBZ);
+               } else if ((technicalGrid.get(i,j,k+1)->SOLVE & compute::BZ) == compute::BZ) {
+                  return bGrid->get(i,j,k+1)->at(fsgrids::bfield::PERBZ);
+               } else {
+                  Real retval = 0.0;
+                  uint nCells = 0;
+                  if ((technicalGrid.get(i-1,j,k)->SOLVE & compute::BZ) == compute::BZ) {
+                     retval += bGrid->get(i-1,j,k)->at(fsgrids::bfield::PERBZ);
+                     nCells++;
+                  }
+                  if ((technicalGrid.get(i+1,j,k)->SOLVE & compute::BZ) == compute::BZ) {
+                     retval += bGrid->get(i+1,j,k)->at(fsgrids::bfield::PERBZ);
+                     nCells++;
+                  }
+                  if ((technicalGrid.get(i,j-1,k)->SOLVE & compute::BZ) == compute::BZ) {
+                     retval += bGrid->get(i,j-1,k)->at(fsgrids::bfield::PERBZ);
+                     nCells++;
+                  }
+                  if ((technicalGrid.get(i,j+1,k)->SOLVE & compute::BZ) == compute::BZ) {
+                     retval += bGrid->get(i,j+1,k)->at(fsgrids::bfield::PERBZ);
+                     nCells++;
+                  }
+                  if (nCells == 0) {
+                     for (int a=i-1; a<i+2; a++) {
+                        for (int b=j-1; b<j+2; b++) {
+                           for (int c=k-1; c<k+2; c++) {
+                              if ((technicalGrid.get(a,b,c)->SOLVE & compute::BZ) == compute::BZ) {
+                                 retval += bGrid->get(a,b,c)->at(fsgrids::bfield::PERBZ);
+                                 nCells++;
+                              }
+                           }
+                        }
+                     }
+                  }
+                  if (nCells == 0) {
+                     cerr << __FILE__ << ":" << __LINE__ << ": ERROR: this should not have fallen through." << endl;
+                     return 0.0;
+                  }
+                  return retval / nCells;
+               }
+            default:
+               cerr << "ERROR: ionosphere boundary tried to copy nonsensical magnetic field component " << component << endl;
+               return 0.0;
+         }
+      } else { // L2 cells
+         Real retval = 0.0;
+         uint nCells = 0;
+         for (int a=i-1; a<i+2; a++) {
```

Comment:
I think this bit can go outside the grid, as there isn't a check first if i,j,k is zero or nx-1,ny-1,nz-1?

Reply:
Now we kind of require the ionosphere not to touch the outer walls. I'll look through again as I solved this issue in another place using the global size of the fsgrid.

Reply:
Oh right I forgot we're in ionosphere and not general sysboundaries :)

Reply:
I can add a check against the global size as in sysboundary.cpp when assigning the bit field, if you want. In practice we never had the ionosphere croos an outer wall in production, only in testing.